src/S3IteratorLibsvm.cpp

```diff
@@ -0,0 +1,412 @@
+  index = index_fw;
+  return result;
+}
+
+bool S3IteratorLibsvm::buildDatasetCsv(
+    const std::string& data,
+    uint64_t index,
+    std::shared_ptr<SparseDataset>& minibatch) {
+  return false;
+}
+
+bool S3IteratorLibsvm::buildDatasetVowpalWabbit(
+    const std::string& data,
+    uint64_t index,
+    std::shared_ptr<SparseDataset>& minibatch) {
+  return false;
+}
+
+/**
+ * Build minibatch from text in libsvm format
+ * We assume index is at a start of a line
+ * index moves forward
+ */
+bool S3IteratorLibsvm::buildDatasetLibsvm(
+    std::string& data,
+    uint64_t& index,
+    std::shared_ptr<SparseDataset>& minibatch) {
+#ifdef DEBUG
+  std::cout << "building dataset libsvm index: " << index << std::endl;
+#endif
+  // libsvm format
+  // <label> <index1>:<value1> <index2>:<value2>
+
+  try {
+    std::vector<std::vector<std::pair<int, FEATURE_TYPE>>> samples;
+    std::vector<FEATURE_TYPE> labels;
+
+    samples.resize(minibatch_rows);
+    labels.resize(minibatch_rows);
+
+    for (uint64_t sample = 0; sample < minibatch_rows; ++sample) {
+      // ignore spaces
+      if (!ignoreSpaces(index, data)) {
+#ifdef DEBUG
+        std::cout << "ignoreSpaces did not find dig index: " << index
+                  << std::endl;
+#endif
+        // did not end up in a digit
+        return false;
+      }
+      int label = readNum<int>(index, data);
+#ifdef DEBUG
+      // std::cout << "read label: " << label << std::endl;
+#endif
+
+      // read pairs
+      while (1) {
+#ifdef DEBUG
+        std::cout << "index: " << index << " reading new pair " << std::endl;
+#endif
+        if (!ignoreSpacesNotNewline(index, data)) {
+          if (data[index] == '\n') {
+#ifdef DEBUG
+            std::cout << "read line index: " << index << std::endl;
+#endif
+            break;  // move to next sample
+          } else if (data[index] == 0) {
+#ifdef DEBUG
+            std::cout << "end of text index: " << index << std::endl;
+#endif
+            return false;  // end of text
+          } else
+            throw std::runtime_error("Error parsing while reading pairs");
+        }
+        uint64_t ind = readNum<uint64_t>(index, data);
+#ifdef DEBUG
+        // std::cout << "index: " << index << " ind: " << ind << std::endl;
+#endif
+        if (data[index] != ':') {
+          return false;
+        }
+        index++;
+
+#ifdef DEBUG
+        std::string num_data = data.substr(index, 10);
+        // std::cout << "reading value: " << num_data << std::endl;
+#endif
+
+        FEATURE_TYPE value = readNum<FEATURE_TYPE>(index, data);
+#ifdef DEBUG
+        // std::cout << "index: " << index << " value: " << value << std::endl;
+#endif
+
+        samples[sample].push_back(std::make_pair(ind, value));
+      }
+      labels[sample] = label;
+    }
+
+#ifdef DEBUG
+    std::cout << "new minibatch" << std::endl;
+#endif
+    minibatch.reset(new SparseDataset(std::move(samples), std::move(labels)));
+    return true;
+  } catch (...) {
```

Comment:
If readNum can't find a digit, there might be something wrong with the dataset (i.e a typo in the dataset). I would add an if statement in this catch statement checking if index is within the size of data to see if we have reached the end of the data. In this case this function would return false, and if not, the function would throw an std::runtime_error.